cmd/specdiff/differ_test.go

```diff
@@ -0,0 +1,282 @@
+)
+
+type testStruct struct {
+	Name    string            `json:"name,omitempty"`
+	TestMap map[string]string `json:"testmap,omitempty"`
+}
+
+func TestDiffProtoStruct(t *testing.T) {
+	tests := []struct {
+		desc         string
+		baseSpec     string
+		revisionSpec string
+		wantProto    *rpc.Diff
+	}{
+		{
+			desc:         "Struct Diff Added",
+			baseSpec:     "./test-specs/base-test.yaml",
+			revisionSpec: "./test-specs/struct-test-add.yaml",
+			wantProto: &rpc.Diff{
+				Added: []string{
+					"components.schemas.Pet.required.age",
+					"components.schemas.Pet.properties.age",
+				},
+				Deleted:      []string{},
+				Modification: map[string]*rpc.Diff_ValueChange{},
+			},
+		},
+		{
+			desc:         "Struct Diff Deleted",
+			baseSpec:     "./test-specs/base-test.yaml",
+			revisionSpec: "./test-specs/struct-test-delete.yaml",
+			wantProto: &rpc.Diff{
+				Added: []string{},
+				Deleted: []string{
+					"components.schemas.Pet.required.name",
+				},
+				Modification: map[string]*rpc.Diff_ValueChange{},
+			},
+		},
+		{
+			desc:         "Struct Diff Modified",
+			baseSpec:     "./test-specs/base-test.yaml",
+			revisionSpec: "./test-specs/struct-test-modify.yaml",
+			wantProto: &rpc.Diff{
+				Added:   []string{},
+				Deleted: []string{},
+				Modification: map[string]*rpc.Diff_ValueChange{
+					"info.version": {
+						To:   "1.0.1",
+						From: "1.0.0",
+					},
+					"components.schemas.Pet.properties.tag.type": {
+						To:   "integer",
+						From: "string",
+					},
+					"components.schemas.Pet.properties.tag.format": {
+						To: "int64",
+					},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			yamlFile, err := ioutil.ReadFile(test.baseSpec)
+			if err != nil {
+				t.Fatalf("Failed to get base spec yaml: %s", err)
+			}
+			yamlFile2, err := ioutil.ReadFile(test.revisionSpec)
+			if err != nil {
+				t.Fatalf("Failed to get revision spec yaml: %s", err)
+			}
+			diffProto, err := GetDiff(yamlFile, yamlFile2)
+			if err != nil {
+				t.Fatalf("Failed to get diff.Diff: %s", err)
+			}
+			opts := cmp.Options{
+				protocmp.Transform(),
+				cmpopts.SortSlices(func(a, b string) bool { return a < b }),
+			}
+			if !cmp.Equal(test.wantProto, diffProto, opts) {
+				t.Errorf("Test %+v returned unexpected diff (-want +got):\n%s", test.desc, cmp.Diff(test.wantProto, diffProto, opts))
+			}
+		})
+	}
+}
+
+func TestMaps(t *testing.T) {
+	tests := []struct {
+		desc      string
+		testMap   reflect.Value
+		change    change
+		wantProto *rpc.Diff
+	}{
+		{
+			desc: "Map Diff String",
+			testMap: reflect.ValueOf(map[string]string{
+				"input1": "result1",
+				"input2": "result2",
+			}),
+			change: change{
+				fieldPath:  []string{},
+				changeType: "added",
+			},
+			wantProto: &rpc.Diff{
+				Added: []string{
+					"input1.result1",
+					"input2.result2",
+				},
+			},
+		}, {
+			desc: "Map Diff Recursive Key",
```

Comment:
I think "Struct Key" is a better description of it. Recursion is an implementation detail.